File: server/php/src/Presentation/Mappers/PaymentMapper.php
<?php

namespace OnlinePayments\ExampleApp\Presentation\Mappers;

use OnlinePayments\ExampleApp\Application\Domain\Enums\Common\Currency;
use OnlinePayments\ExampleApp\Application\Domain\Enums\Payments\PaymentMethodType;
use OnlinePayments\ExampleApp\Application\DTOs\Payment\RequestDto;
use OnlinePayments\ExampleApp\Application\DTOs\Payment\ResponseDto;
use OnlinePayments\ExampleApp\Presentation\Models\Payment\Request;
use OnlinePayments\ExampleApp\Presentation\Models\Payment\Response;

final class PaymentMapper
{
    private const SMALLEST_UNIT = 100;

    public static function map(Request $request): RequestDto
    {
        return new RequestDto(
            amount: (int) ($request->amount * self::SMALLEST_UNIT),
            currency: Currency::from($request->currency),
            method: PaymentMethodType::from($request->method),
            hostedTokenizationId: $request->hostedTokenizationId,
            shippingAddress: $request->shippingAddress ?? null,
            billingAddress: $request->billingAddress ?? null,
            card: $request->card ?? null,
            mandate: $request->mandate ?? null
        );
    }

    public static function mapResponse(ResponseDto $responseDto): Response
    {
        return new Response(
            $responseDto->id ?? null,
            $responseDto->status ?? null,
            $responseDto->statusOutput ?? null
        );
    }
}

File: server/python/app/presentation/mappers/payment_mapper.py
from app.application.domain.enums.common.currency import Currency
from app.application.domain.enums.payments.payment_method_type import PaymentMethodType
from app.application.domain.payments.status_output import StatusOutput
from app.application.dtos.create_payment.request_dto import CreatePaymentRequestDto
from app.application.dtos.create_payment.response_dto import CreatePaymentResponseDto
from app.presentation.models.create_payment.request import CreatePaymentRequest
from app.presentation.models.create_payment.response import CreatePaymentResponse

SMALLEST_UNIT = 100

class PaymentPresentationMapper:

    @staticmethod
    def to_dto(request: CreatePaymentRequest) -> CreatePaymentRequestDto:
        return CreatePaymentRequestDto(
            amount=int(float(request.amount) * SMALLEST_UNIT),
            currency=Currency(request.currency) if request.currency else None,
            method=PaymentMethodType(request.method) if request.method else None,
            hosted_tokenization_id=request.hosted_tokenization_id,
            billing_address=request.billing_address,
            shipping_address=request.shipping_address,
            card=request.card,
            mandate=request.mandate
        )

    @staticmethod
    def from_dto(response_dto: CreatePaymentResponseDto) -> CreatePaymentResponse:
        status_output = None
        if response_dto.status_output:
            status_output = StatusOutput(
                status_code=response_dto.status_output.status_code,
                status_category=response_dto.status_output.status_category
            )
        return CreatePaymentResponse(
            id=response_dto.id,
            status=response_dto.status,
            status_output=status_output
        )